File: src/main/java/com/pesocial/service/StoryService.java
package com.pesocial.service;

import com.pesocial.dto.StoryDetailDTO;
import com.pesocial.model.Story;
import java.util.List;

public interface StoryService {
  Story createStory(String authorId, String mediaUrl);

  Story getStoryById(String storyId);

  List<StoryDetailDTO> getActiveStories(String currentUserId);

  void viewStory(String storyId, String userId);

  void likeStory(String storyId, String userId);

  void unlikeStory(String storyId, String userId);

  StoryDetailDTO getStoryAnalytics(String storyId, String currentUserId);

  List<StoryDetailDTO> getUserStories(String userId, String currentUserId);

  void deleteStory(String storyId);
}

File: pesocial-frontend/src/services/storyService.ts
import api from './api'
import type { Story, StoryDetailDTO } from '../types/story'

class StoryService {
  async createStory(mediaUrl: string): Promise<Story> {
    const { data } = await api.post<Story>('/api/stories', { mediaUrl })
    return data
  }

  async getActiveStories(): Promise<StoryDetailDTO[]> {
    const { data } = await api.get<StoryDetailDTO[]>('/api/stories/active')
    return data
  }

  async getUserStories(userId: string): Promise<StoryDetailDTO[]> {
    const { data } = await api.get<StoryDetailDTO[]>(`/api/stories/user/${encodeURIComponent(userId)}`)
    return data
  }

  async viewStory(storyId: string): Promise<void> {
    await api.post(`/api/stories/${encodeURIComponent(storyId)}/view`)
  }

  async likeStory(storyId: string): Promise<void> {
    await api.post(`/api/stories/${encodeURIComponent(storyId)}/like`)
  }

  async unlikeStory(storyId: string): Promise<void> {
    await api.post(`/api/stories/${encodeURIComponent(storyId)}/unlike`)
  }

  async getStoryAnalytics(storyId: string): Promise<StoryDetailDTO> {
    const { data } = await api.get<StoryDetailDTO>(`/api/stories/${encodeURIComponent(storyId)}/analytics`)
    return data
  }

  async deleteStory(storyId: string): Promise<void> {
    await api.delete(`/api/stories/${encodeURIComponent(storyId)}`)
  }

  async uploadStoryMedia(file: File): Promise<string> {
    const formData = new FormData()
    formData.append('file', file)

    // Reuse proven GridFS upload flow already used by chat attachments.
    const { data } = await api.post<{ attachmentUrl: string }>('/api/messages/attachments', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })

    return data.attachmentUrl
  }
}

export const storyService = new StoryService()
